build-system/compile/json-schema/index.js

```diff
@@ -0,0 +1,121 @@
+ *   validate("my-data");
+ *   ```
+ */
+const {default: Ajv} = require('ajv');
+const dedent = require('dedent');
+const {outputFile, pathExists} = require('fs-extra');
+const {basename, join, relative} = require('path');
+const {once} = require('../../common/once');
+const {
+  TransformCache,
+  batchedRead,
+  md5,
+} = require('../../common/transform-cache');
+
+/**
+ * File filter to match import.
+ * It's useful that this always ends in .schema.json, since that suffix enables
+ * JSON Schema features on IDEs like VSCode.
+ * @type {RegExp}
+ */
+const importFilter = /\.schema\.json$/;
+
+// https://ajv.js.org/options.html
+const ajvOptions = {
+  code: {
+    esm: true,
+    source: true,
+    lines: true,
+  },
+  allErrors: true,
+  validateSchema: false,
+};
+
+const ajv = once(() => new Ajv(ajvOptions));
+
+const escapeJsIdentifier = (id) => id.replace(/[^a-z_0-9]/gi, '_');
+
+const getTransformCache = once(() => new TransformCache('.json-schema-cache'));
+
+/**
+ * Like batchRead(), but considers ajvOptions part of the invalidation hash.
+ * @param {string} filename
+ * @return {Promise<{contents: string, hash: string}>}
+ */
+async function ajvBatchedRead(filename) {
+  const {contents, hash} = await batchedRead(filename);
+  const rehash = md5(JSON.stringify({ajvOptions, hash}));
+  return {contents, hash: rehash};
+}
+
+/**
+ * @param {string} filename
+ * @return {Promise<string>}
+ */
+async function getCompiledJsonSchemaFilename(filename) {
+  const {contents, hash} = await ajvBatchedRead(filename);
+  const transformCache = getTransformCache();
+  const cached = await transformCache.get(hash);
+  let output = cached;
+  if (!output) {
+    output = avjCompile(filename, contents);
+    transformCache.set(hash, Promise.resolve(output));
+  }
+  const outputFilename = `build/${filename}.js`;
+  if (!cached || !(await pathExists(outputFilename))) {
+    await outputFile(outputFilename, output);
+  }
+  return outputFilename;
+}
+
+/**
+ * @param {string} filename
+ * @param {string} contents
+ * @return {string}
+ */
+function avjCompile(filename, contents) {
+  const schema = JSON.parse(contents);
+  const validateAjv = ajv().compile(schema);
+  const validateFnName = escapeJsIdentifier(
+    `validate_${basename(filename, '.json')}`
+  );
+  return (
+    dedent(`
+      const validateAjv = __validateAjv__;
+      /**
+       * @param {any} data
+       * @return {{params?: object, message?: string, data?: any}[]}
+       * https://ajv.js.org/api.html#error-objects
+       */
+      export default function __validateFnName__(data) {
+        return validateAjv(data) ? [] : [...validateAjv.errors]
+      }
+    `)
+      .replace('__validateFnName__', validateFnName)
+      // validateAjv returns a boolean, and modifies a property of the function
+      // for errors. The default instead returns an array of errors, which is
+      // empty when the input is valid.
+      .replace('__validateAjv__', validateAjv.toString())
```

Comment:
AJV seems to [support](https://ajv.js.org/standalone.html#generating-functions-s-for-a-single-schema-using-the-js-library-es6-and-cjs-exports) generating a validation library via the `ajv/dist/standalone` package:

```js
const ajv = new Ajv({ code: { source: true, esm: true } });
const validate = ajv.compile(schema);
return standaloneCode(ajv, validate);
```

This is considerably safer than depending on the `toString` of a function, as it is not guaranteed to be runnable code (you can't serialize the closure's scope, for example).

Reply:
We need to wrap the generated function since the result interface is somewhat awkward (see comment). `standaloneCode` forbids us from doing this.

However, the result of `ajv.compile` contains a `source.validateCode` field that's safe and fits our purposes.